# This Notebook enables you to call every Function in the MTurk Functions catalog

### Prerequisites:
1. You have [an MTurk Requester account and linked AWS account](https://medium.com/@mechanicalturk/setting-up-your-amazon-mechanical-turk-mturk-requester-account-and-aws-account-a316e2f6a156)
2. You have [completed the account set up steps to call AmazonMechanicalTurk and AmazonMechanicalTurkCrowd](https://medium.com/@mechanicalturk/setting-up-your-amazon-mechanical-turk-and-aws-accounts-to-call-a-preview-api-59ade8beafd1)
3. You have configured a local AWS profile to call MTurk with the user you set up in step 2
4. You have [prepaid for HITs in your MTurk Requester account](https://requester.mturk.com/prepayments/new) to cover paying for Worker rewards (suggested amount to start is $5.00)

### If you haven't done so, install the mturk-crowd-beta Python client

In [ ]:
!pip install --upgrade mturk-crowd-beta-client --ignore-installed six

## Next we will import the packages we need to use the API
You'll need to do this before calling any API

In [1]:
from mturk_crowd_beta_client import MTurkCrowdClient
from boto3.session import Session
import uuid

## Next we set up a session

This examples assume you have a local AWS profile named __'mturk-crowd-caller'__, but you can authenticate however you like, including by directly passing in your access key and secret key.  The access key and secret key are for the User you created during setup with the policies AmazonMechanicalTurkFullAccess and AmazonMechanicalTurkCrowdFullAccess.

In [2]:
#session = Session(profile_name='mturk-crowd-caller')
session = Session(profile_name='mtf-caller')

crowd_client = MTurkCrowdClient(session)

## Available APIs:

### Natural Language Processing

* [sentiment-analysis](#sentiment-analysis)
* [emotion-detection](#emotion-detection)
* [semantic-similarity](#semantic-similarity)
* [collect-utterance-text](#collect-utterance-text)
* [text-intent-detection](#text-intent-detection)
* [text-categorization](#text-categorization)
* [named-entity-recognition](#named-entity-recognition)
* [co-reference-resolution](#co-reference-resolution)
* [key-phrase-extraction](#key-phrase-extraction)

### Computer Vision
* [image-contains](#image-contains)
* [image-categorization](#image-categorization)
* [image-similarity](#image-similarity)
* [bounding-box](#bounding-box)
* [face-detection](#face-detection) (not public)
* [text-recognition](#text-recognition) (not public)
* [logo-recognition](#logo-recognition) (not public)


## <a id='sentiment-analysis'>sentiment-analysis</a>

Input:
{
 "text": "Everything is wonderful!"
}

Result: {'sentiment': 'positive'}

Max length of the input text is 400 characters.  Sentiment is one of positive, negative, neutral or cannot determine.  We ask up to 5 Workers until 2 of them agree on the answer.


When you create a Task using the sentiment-analysis API, you're automatically creating a Human Intelligence Task (HIT) on worker.mturk.com.  Here's an example of a sentiment analysis HIT.

![sentiment-analysis-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/sentiment-analysis-HIT-example.png)

### Create a Task

In [ ]:
#set the function_name to the name of the API
function_name = 'sentiment-analysis'

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# define the text that you want analyzed, up to 16k
text = 'The trip by @VP Pence was long planned. He is receiving great praise for leaving game after the players showed such disrespect for country!'

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'text': text})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

### Creating multiple Tasks and Processing Results 

Next, we'll read from a CSV file that has multiple rows of text that needs to be analyzed for sentiment using the pandas library.  As a prerequisite you need to have installed the pandas library.

In [ ]:
import pandas as pd

In [ ]:
#read the input data from csv into a pandas DataFrame
data = pd.read_csv('https://s3-us-west-2.amazonaws.com/mturk-sample-datasets/sentiment-analysis-example-inputs.csv')

In [ ]:
# column 1 is the Task ID and column2 is the input data
data

In [ ]:
#loop through the rows and create a Task per row, write the put_task status in a new column
status_codes = []
for index, row in data.iterrows():
    put_result = crowd_client.put_task(function_name, row.task_id,{'text': row.review_text})
    status_codes.append(put_result.status_code)
data['status_codes'] = status_codes

In [ ]:
data

We'll call get_task periodically until all of the Tasks reach the "completed" state.

In [ ]:
# loop through the rows and get results, store the state and results in new columns

status_codes = []
responses = []
for index, row in data.iterrows():
    get_result = crowd_client.get_task(function_name, row.task_id)
    status_codes.append(get_result.status_code)
    responses.append(get_result.json())

In [ ]:
data['status_codes'] = status_codes
data['state'] = [r['state'] for r in responses]
data['problem_details'] = [r['problemDetails'] for r in responses]
#need to handle the case when not all tasks are in completed state
data['sentiment'] = [r['result']['sentiment'] for r in responses]

In [ ]:
data

## <a id='image-contains'>image-contains</a>

Given an image and a type of object to look for, this API returns true or false.

Input: `{"image": {"url": "https://www.mturk.com/media/butterbean.jpg"}, "target": {"label": "dog"} }`

Result: `{"containsTarget": true}`

Internal Only: we ask 2 - 5 Workers to get agreement.

In [48]:
function_name = 'image-contains'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# the URL of the image that you want annotated
image_url = 'https://urbanedge.blogs.rice.edu/files/2016/02/midtown-15wd4ck.jpg'

### The type of thing we're looking for

In [ ]:
label = 'Pedestrians'

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                                   task_name,
                                   {'image': {'url': image_url}, 'target': {'label': label}})

print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))        

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(
{'status_code': get_result.status_code, 'task': get_result.json()}))

### Creating multiple Tasks and Processing Results 

Next, we'll read from a CSV file that has multiple rows of text that needs to be analyzed for sentiment using the pandas library.  As a prerequisite you need to have installed the pandas library.

In [ ]:
import pandas as pd

In [49]:
#read the input data from csv into a pandas DataFrame
data = pd.read_csv('https://s3-us-west-2.amazonaws.com/mturk-sample-datasets/image-contains-example-inputs.csv')

In [50]:
# column 1 is the Task ID and column2 amd 3 are the input data
data

task_id                                          image_url  \
0   image-contains-1  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
1   image-contains-2  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
2   image-contains-3  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
3   image-contains-4  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
4   image-contains-5  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
5   image-contains-6  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
6   image-contains-7  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
7   image-contains-8  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
8   image-contains-9  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
9  image-contains-10  https://s3-us-west-2.amazonaws.com/mturk-sampl...   

                 Label  
0       womens satchel  
1    womens flip flops  
2  womens ballet flats  
3   womens wedge shoes  
4    womens flip flops  
5   womens wedge shoes  
6         womens mules  
7         womens mules  
8   womens ankle boots  
9   womens wedge shoes

In [51]:
#loop through the rows and create a Task per row, write the put_task status in a new column
status_codes = []
for index, row in data.iterrows():
    put_result = crowd_client.put_task(function_name, 
                                       row.task_id,
                                       {'image': {'url': row.image_url},
                                        'target': {'label': row.Label} })
    status_codes.append(put_result.status_code)
data['status_codes'] = status_codes

In [52]:
data

task_id                                          image_url  \
0   image-contains-1  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
1   image-contains-2  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
2   image-contains-3  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
3   image-contains-4  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
4   image-contains-5  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
5   image-contains-6  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
6   image-contains-7  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
7   image-contains-8  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
8   image-contains-9  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
9  image-contains-10  https://s3-us-west-2.amazonaws.com/mturk-sampl...   

                 Label  status_codes  
0       womens satchel           200  
1    womens flip flops           200  
2  womens ballet flats           200  
3   womens wedge shoes           200  
4    womens flip flops           200  
5   womens wedge shoes           200  
6         womens mules           200  
7         womens mules           200  
8   womens ankle boots           200  
9   womens wedge shoes           200

We'll call get_task periodically until all of the Tasks reach the "completed" state.

In [53]:
# loop through the rows and get results, store the state and results in new columns

status_codes = []
responses = []
for index, row in data.iterrows():
    get_result = crowd_client.get_task(function_name, row.task_id)
    status_codes.append(get_result.status_code)
    responses.append(get_result.json())

In [54]:
data['status_codes'] = status_codes
data['state'] = [r['state'] for r in responses]
data['problem_details'] = [r['problemDetails'] for r in responses]
#need to handle the case when not all tasks are in completed state
data['containsTarget'] = [r['result']['containsTarget'] for r in responses]

In [55]:
data

task_id                                          image_url  \
0   image-contains-1  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
1   image-contains-2  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
2   image-contains-3  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
3   image-contains-4  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
4   image-contains-5  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
5   image-contains-6  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
6   image-contains-7  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
7   image-contains-8  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
8   image-contains-9  https://s3-us-west-2.amazonaws.com/mturk-sampl...   
9  image-contains-10  https://s3-us-west-2.amazonaws.com/mturk-sampl...   

                 Label  status_codes      state problem_details  \
0       womens satchel           200  completed            None   
1    womens flip flops           200  completed            None   
2  womens ballet flats           200  completed            None   
3   womens wedge shoes           200  completed            None   
4    womens flip flops           200  completed            None   
5   womens wedge shoes           200  completed            None   
6         womens mules           200  completed            None   
7         womens mules           200  completed            None   
8   womens ankle boots           200  completed            None   
9   womens wedge shoes           200  completed            None   

   containsTarget  
0           False  
1           False  
2            True  
3           False  
4           False  
5            True  
6            True  
7            True  
8            True  
9            True

In [57]:
data.to_csv('image-contains-results.csv')

## <a id='emotion-detection'>emotion-detection</a>

This API determines the emotion of text.

Input:{"text": "First time ever winning all three fantasy leagues AND @Seahawks win!"}
Result: {"emotion": "joy"}

Emotion is one of Joy, Anger, Fear, Sadness, Surprise, Disgust, or Neutral.  Internal only: we ask up to 9 Workers to get agreement.

![emotion-detection-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/emotion-detection-HIT-example.png)

In [ ]:
function_name = 'emotion-detection'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# define the text that you want analyzed, up to 16k
text = 'Just realised that I have #280characters so now I can finally tweet this: "Daenerys Stormborn of the House Targaryen, First of Her Name, the Unburnt, Queen of the Andals and the First Men, Khaleesi of the Great Grass Sea, Breaker of Chains, and Mother of Dragons'

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'text': text})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id=semantic-similarity>semantic-similarity</a>

Compare two text documents and rate them on how similar they are, on a scale between 0 and 1 where 1 is very similar.

Input: {"text1": "The sky is blue.", "text2": "The sky was the color of blue."}

Result: {"similarityScore": 0.75}

In [ ]:
function_name = 'semantic-similarity'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# define the text that you want analyzed, up to 16k
text1 = 'I\'m so hungry I could eat a horse'
text2 = 'I\'m hangry'

In [ ]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'text1': text1, 'text2': text2})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

### Creating multiple Tasks and Processing Results 

Next, we'll read from a CSV file that has multiple rows of text that needs to be analyzed for sentiment using the pandas library.  As a prerequisite you need to have installed the pandas library.

In [ ]:
import pandas as pd

In [ ]:
#read the input data from csv into a pandas DataFrame
data = pd.read_csv('https://s3-us-west-2.amazonaws.com/mturk-sample-datasets/semantic-similarity-example-inputs.csv')

In [ ]:
# column 1 is the Task ID and column2 is the input data
data

In [ ]:
#loop through the rows and create a Task per row, write the put_task status in a new column
status_codes = []
for index, row in data.iterrows():
    put_result = crowd_client.put_task(function_name, 
                                       row.task_id,
                                        {'text1': row.review1, 'text2': row.review2})
    status_codes.append(put_result.status_code)
data['status_codes'] = status_codes

In [ ]:
data

We'll call get_task periodically until all of the Tasks reach the "completed" state.

In [ ]:
# loop through the rows and get results, store the state and results in new columns

status_codes = []
responses = []
for index, row in data.iterrows():
    get_result = crowd_client.get_task(function_name, row.task_id)
    status_codes.append(get_result.status_code)
    responses.append(get_result.json())

In [ ]:
data['status_codes'] = status_codes
data['state'] = [r['state'] for r in responses]
data['problem_details'] = [r['problemDetails'] for r in responses]
#need to handle the case when not all tasks are in completed state
data['similarityScore'] = [r['result']['similarityScore'] for r in responses]

In [ ]:
data

## <a id='collect-utterance-text'>collect-utterance-text</a>

Given a context and an intention, provide what you would say, in text, in that situation.

input: {"context": "Someone recently bought a phone and it doesn't work", "intent": "They want to return a phone"}'

result: {'utterances': [{'text': "result: {"utterance": "The phone that I just bought stopped working.  I want to get a refund"}
"}]}

Internal only: We ask only 1 Worker for each task.

Here's an example of what the HIT might look like to Workers:
![collect-utterance-text-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/collect-utterance-text-HIT-example.png)

In [ ]:
function_name = 'collect-utterance-text'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# provide the context and the intent
context = 'Someone is calling their doctor\'s office'
intent = 'I want to change an existing appointment'

In [ ]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'context': context, 'intent': intent})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id='image-categorization'>image-categorization</a>

Given an image URL and a list of categories, determine which category the image best belongs to.  You can optionally provide a description for each category to provide more guidance on what each category includes.

Input: {"image":{"url": "https://requester.mturk.com/assets/simon.jpg"},"categories": “categories”: [{“label”: “Plant”}, {“label”:”Animal”}, {“label”:”Bacteria”}, {“label”:”Fungi”}, {“label”:”Protists”}]

Result: {"category":"Animal"} 

![image-categorization-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/image-categorization-HIT-example.png)

In [ ]:
function_name = 'image-categorization'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# provide the context and the intent
image_url = 'https://www.abebooks.com/images/books/harry-potter/sorcerers-stone.jpg'
categories =  [{'label': 'Hardcover', 'description': 'hardcover version'}, {'label':'Paperback', 'description': 'paperback version'}, {'label':'eBook', 'description': 'digital text version, including PDF, Kindle'}, {'label':'Audio Book', 'description': 'any audio version, including CD, mp3, streaming'}]

In [ ]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                            {'image': {'url': image_url},
                            'categories': categories})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id='text-intent-detection'>text-intent-detection</a>

Given a list of intentions, categorize the best intention that matches the text given.

input: {"text": "foo", "intents": [{"label": "Schedule an appointment", "description": "example: I need to make an appointment with Dr. Smith"}, {"label": "Medical Record Request", "description": "I need a copy of my kids&#39; immunication records"}]}

result:{'applicableIntents': [{'label': 'Schedule an appointment'}]}

Here's an example of what the task might look like for Workers.
![text-intent-detection-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/text-intent-detection-HIT-example.png)

In [ ]:
function_name = 'text-intent-detection'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# define the text to be analysed and a list of intentions to categorize the text for
text = 'my son was stung by bees and I need to know if I need to go to the ER'
intents =  [{'label': 'Schedule an appointment', 'description': 'e.g. I need to make an appointment with Dr. Smith'}, {'label': 'Medical Record Request', 'description': 'e.g. I need a copy of my kids immunization records'}]

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                                   task_name,
                                  {'text': text,
                                  'intents': intents})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id=text-categorization>text-categorization</a>

Given a piece of text and a list of categories, choose the best cateogry that fits the text.

input: {"text": "The Baltimore Ravens beat the Seattle Seahawks 24-7", "categories": [{"label": "sports", "description": "talks about sports"}, {"label": "food", "description": "talks about food"}]}

result: {'applicableCategories': [{'label': 'sports'}]}

Here's an example of a HIT to MTurk Workers generated by calling this API
![text-categorization-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/text-categorization-HIT-example.png)

In [ ]:
function_name = 'text-categorization'

### Create a Task

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# define the text to be analysed and a list of categories to assess for
text = 'These are great. They do run a touch small. I almost could go a half size up from my normal size.'
categories =  [{'label': 'style', 'description': 'related to the look of the product'}, {'label': 'fit', 'description': 'related to the sizing or how it fits'}, {'label': 'quality', 'description': 'related to how well made the product is'}, {'label': 'price', 'description': 'related to cost or value of the product'}]

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                                   task_name,
                                  {'text': text,
                                  'categories': categories})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id=named-entity-recognition>named-entity-recognition</a>

Given a paragraph of text, label the span of text that represent entities of various categories, such as person and location.

input: {'text': 'Harry Potter is midway through his training as a wizard and his coming of age. Harry wants to get away from the pernicious Dursleys and go to the International Quidditch Cup. He wants to find out about the mysterious event that's supposed to take place at Hogwarts this year, an event involving two other rival schools of magic, and a competition that hasn't happened for a hundred years. He wants to be a normal, fourteen-year-old wizard. But unfortunately for Harry Potter, he's not normal - even by wizarding standards. And in his case, different can be deadly.'}

result: {'entities': [{'end': 12, 'start': 0, 'text': 'Harry Potter', 'type': 'person'}, {'end': 84, 'start': 79, 'text': 'Harry', 'type': 'person'}, {'end': 131, 'start': 123, 'text': 'Dursleys', 'type': 'person'}, {'end': 77, 'start': 60, 'text': 'his coming of age', 'type': 'date'}, {'end': 173, 'start': 146, 'text': 'International Quidditch Cup', 'type': 'event'}, {'end': 222, 'start': 217, 'text': 'event', 'type': 'event'}, {'end': 284, 'start': 279, 'text': 'event', 'type': 'event'}, {'end': 346, 'start': 335, 'text': 'competition', 'type': 'event'}, {'end': 264, 'start': 256, 'text': 'Hogwarts', 'type': 'organization'}, {'end': 318, 'start': 311, 'text': 'schools', 'type': 'organization'}, {'end': 264, 'start': 256, 'text': 'Hogwarts', 'type': 'organization'}]}

![NER-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/ner-HIT-example.png)

### Create a Task

In [ ]:
#set the function_name to the name of the API
function_name = 'named-entity-recognition'

In [ ]:
# automatically generate a random task ID
# task_name = 'my-test-task-' + uuid.uuid4().hex
task_name='my-test-task-ff7d0b50791a43cc819b60b7b568c422'
print(task_name)

In [ ]:
# define the text that you want analyzed, up to 16k
text = 'Harry Potter is midway through his training as a wizard and his coming of age. Harry wants to get away from the pernicious Dursleys and go to the International Quidditch Cup. He wants to find out about the mysterious event that\'s supposed to take place at Hogwarts this year, an event involving two other rival schools of magic, and a competition that hasn\'t happened for a hundred years.'

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'text': text})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id='bounding-box'> bounding-box </a>

Given an image, draw a box around particular objects, specified by labels.  This API returns the coordinates of labeled objects in the image.

input: {"image": {"url": "https://sierra-videos.s3.amazonaws.com/2017_0521/DJI_0001_frames_720h/1495411174030000000.jpg"}, "labels": [{"label": "car"}, {"label": "pedestrians"}]}

result: {'labels': [{'boundingBox': {'height': 55, 'left': 896, 'top': 665, 'width': 51}, 'label': 'car'}, {'boundingBox': {'height': 50, 'left': 555, 'top': 515, 'width': 63}, 'label': 'car'}, {'boundingBox': {'height': 51, 'left': 539, 'top': 590, 'width': 58}, 'label': 'car'}, {'boundingBox': {'height': 44, 'left': 216, 'top': 512, 'width': 55}, 'label': 'car'}, {'boundingBox': {'height': 54, 'left': 873, 'top': 475, 'width': 53}, 'label': 'car'}, {'boundingBox': {'height': 56, 'left': 1083, 'top': 490, 'width': 47}, 'label': 'car'}, {'boundingBox': {'height': 42, 'left': 164, 'top': 659, 'width': 66}, 'label': 'car'}, {'boundingBox': {'height': 46, 'left': 524, 'top': 213, 'width': 50}, 'label': 'car'}, {'boundingBox': {'height': 48, 'left': 814, 'top': 311, 'width': 53}, 'label': 'car'}]}

This API currently only supports upright bounding boxes.  

Here's an example of what the Worker sees:
![bound-box-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/bound-box-HIT-example.png)

In [8]:
function_name = 'bounding-box'

### Create a Task

In [9]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

my-test-task-52162080f2964bbeafc7a1b3e42dc78e


In [10]:
# provide the context and the intent
image_url = 'https://sierra-videos.s3.amazonaws.com/2017_0521/DJI_0001_frames_720h/1495411174030000000.jpg'
labels =  [{"label": "car"}, {"label": "pedestrians"}]

In [11]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                            {'image': {'url': image_url},
                            'labels': labels})

print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

PUT response: {'status_code': 201, 'task': {'input': {'image': {'url': 'https://sierra-videos.s3.amazonaws.com/2017_0521/DJI_0001_frames_720h/1495411174030000000.jpg'}, 'labels': [{'label': 'car'}, {'label': 'pedestrians'}]}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-52162080f2964bbeafc7a1b3e42dc78e'}}


### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [12]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

GET response: {'status_code': 200, 'task': {'input': {'image': {'url': 'https://sierra-videos.s3.amazonaws.com/2017_0521/DJI_0001_frames_720h/1495411174030000000.jpg'}, 'labels': [{'label': 'car'}, {'label': 'pedestrians'}]}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-52162080f2964bbeafc7a1b3e42dc78e'}}


### Creating multiple Tasks and Processing Results 

Next, we'll read from a CSV file that has multiple rows of text that needs to be analyzed for sentiment using the pandas library.  As a prerequisite you need to have installed the pandas library.

In [13]:
import pandas as pd

In [14]:
#read the input data from csv into a pandas DataFrame
data = pd.read_csv('https://s3-us-west-2.amazonaws.com/mturk-sample-datasets/vehicle_box_annotation_images-input-sample.csv')

In [15]:
# column 1 is the Task ID and column2 is the image_url and column3 is the item to draw
data

task_id                                          image_url label
0    bounding-box-10  https://sierra-videos.s3.amazonaws.com/2017_05...   car
1    bounding-box-20  https://sierra-videos.s3.amazonaws.com/2017_05...   car
2    bounding-box-30  https://sierra-videos.s3.amazonaws.com/2017_05...   car
3    bounding-box-40  https://sierra-videos.s3.amazonaws.com/2017_05...   car
4    bounding-box-50  https://sierra-videos.s3.amazonaws.com/2017_05...   car
5    bounding-box-60  https://sierra-videos.s3.amazonaws.com/2017_05...   car
6    bounding-box-70  https://sierra-videos.s3.amazonaws.com/2017_05...   car
7    bounding-box-80  https://sierra-videos.s3.amazonaws.com/2017_05...   car
8    bounding-box-90  https://sierra-videos.s3.amazonaws.com/2017_05...   car
9   bounding-box-100  https://sierra-videos.s3.amazonaws.com/2017_05...   car
10  bounding-box-110  https://sierra-videos.s3.amazonaws.com/2017_05...   car
11  bounding-box-120  https://sierra-videos.s3.amazonaws.com/2017_05...   car
12  bounding-box-130  https://sierra-videos.s3.amazonaws.com/2017_05...   car
13  bounding-box-140  https://sierra-videos.s3.amazonaws.com/2017_05...   car
14  bounding-box-150  https://sierra-videos.s3.amazonaws.com/2017_05...   car
15  bounding-box-160  https://sierra-videos.s3.amazonaws.com/2017_05...   car
16  bounding-box-170  https://sierra-videos.s3.amazonaws.com/2017_05...   car
17  bounding-box-180  https://sierra-videos.s3.amazonaws.com/2017_05...   car
18  bounding-box-190  https://sierra-videos.s3.amazonaws.com/2017_05...   car
19  bounding-box-200  https://sierra-videos.s3.amazonaws.com/2017_05...   car
20  bounding-box-210  https://sierra-videos.s3.amazonaws.com/2017_05...   car
21  bounding-box-220  https://sierra-videos.s3.amazonaws.com/2017_05...   car
22  bounding-box-230  https://sierra-videos.s3.amazonaws.com/2017_05...   car
23  bounding-box-240  https://sierra-videos.s3.amazonaws.com/2017_05...   car
24  bounding-box-250  https://sierra-videos.s3.amazonaws.com/2017_05...   car
25  bounding-box-260  https://sierra-videos.s3.amazonaws.com/2017_05...   car
26  bounding-box-270  https://sierra-videos.s3.amazonaws.com/2017_05...   car
27  bounding-box-280  https://sierra-videos.s3.amazonaws.com/2017_05...   car
28  bounding-box-290  https://sierra-videos.s3.amazonaws.com/2017_05...   car
29  bounding-box-300  https://sierra-videos.s3.amazonaws.com/2017_05...   car
..               ...                                                ...   ...
62  bounding-box-630  https://sierra-videos.s3.amazonaws.com/2017_05...   car
63  bounding-box-640  https://sierra-videos.s3.amazonaws.com/2017_05...   car
64  bounding-box-650  https://sierra-videos.s3.amazonaws.com/2017_05...   car
65  bounding-box-660  https://sierra-videos.s3.amazonaws.com/2017_05...   car
66  bounding-box-670  https://sierra-videos.s3.amazonaws.com/2017_05...   car
67  bounding-box-680  https://sierra-videos.s3.amazonaws.com/2017_05...   car
68  bounding-box-690  https://sierra-videos.s3.amazonaws.com/2017_05...   car
69  bounding-box-700  https://sierra-videos.s3.amazonaws.com/2017_05...   car
70  bounding-box-710  https://sierra-videos.s3.amazonaws.com/2017_05...   car
71  bounding-box-720  https://sierra-videos.s3.amazonaws.com/2017_05...   car
72  bounding-box-730  https://sierra-videos.s3.amazonaws.com/2017_05...   car
73  bounding-box-740  https://sierra-videos.s3.amazonaws.com/2017_05...   car
74  bounding-box-750  https://sierra-videos.s3.amazonaws.com/2017_05...   car
75  bounding-box-760  https://sierra-videos.s3.amazonaws.com/2017_05...   car
76  bounding-box-770  https://sierra-videos.s3.amazonaws.com/2017_05...   car
77  bounding-box-780  https://sierra-videos.s3.amazonaws.com/2017_05...   car
78  bounding-box-790  https://sierra-videos.s3.amazonaws.com/2017_05...   car
79  bounding-box-800  https://sierra-videos.s3.amazonaws.com/2017_05...   car
80  bounding-box-810  https://sierra-videos.s3.amazonaws.com/2017_05...   car
81  bounding-box-820  https://sierr

In [16]:
#loop through the rows and create a Task per row, write the put_task status in a new column
status_codes = []
for index, row in data.iterrows():
    put_result = crowd_client.put_task(function_name, 
                                       row.task_id,
                                       {'image': {'url': row.image_url},
                                        'labels' : [{"label": row.label}]})
#    print('PUT response: {}'.format(
#    {'status_code': put_result.status_code, 'task': put_result.json()}))
    status_codes.append(put_result.status_code)
data['status_codes'] = status_codes

In [19]:
data

task_id                                          image_url label  \
0    bounding-box-10  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
1    bounding-box-20  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
2    bounding-box-30  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
3    bounding-box-40  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
4    bounding-box-50  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
5    bounding-box-60  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
6    bounding-box-70  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
7    bounding-box-80  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
8    bounding-box-90  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
9   bounding-box-100  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
10  bounding-box-110  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
11  bounding-box-120  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
12  bounding-box-130  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
13  bounding-box-140  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
14  bounding-box-150  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
15  bounding-box-160  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
16  bounding-box-170  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
17  bounding-box-180  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
18  bounding-box-190  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
19  bounding-box-200  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
20  bounding-box-210  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
21  bounding-box-220  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
22  bounding-box-230  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
23  bounding-box-240  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
24  bounding-box-250  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
25  bounding-box-260  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
26  bounding-box-270  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
27  bounding-box-280  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
28  bounding-box-290  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
29  bounding-box-300  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
..               ...                                                ...   ...   
62  bounding-box-630  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
63  bounding-box-640  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
64  bounding-box-650  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
65  bounding-box-660  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
66  bounding-box-670  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
67  bounding-box-680  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
68  bounding-box-690  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
69  bounding-box-700  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
70  bounding-box-710  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
71  bounding-box-720  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
72  bounding-box-730  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
73  bounding-box-740  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
74  bounding-box-750  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
75  bounding-box-760  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
76  bounding-box-770  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
77  bounding-box-780  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
78  bounding-box-790  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
79  bounding-box-800  https://sierra-videos.

We'll call get_task periodically until all of the Tasks reach the "completed" state.

In [17]:
# loop through the rows and get results, store the state and results in new columns

status_codes = []
responses = []
for index, row in data.iterrows():
    get_result = crowd_client.get_task(function_name, row.task_id)
    status_codes.append(get_result.status_code)
    responses.append(get_result.json())

In [45]:
data['status_codes'] = status_codes
data['state'] = [r['state'] for r in responses]
data['problem_details'] = [r['problemDetails'] for r in responses]
#need to handle the case when not all tasks are in completed state
data['result'] = [r['result'] for r in responses]

In [46]:
data

task_id                                          image_url label  \
0    bounding-box-10  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
1    bounding-box-20  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
2    bounding-box-30  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
3    bounding-box-40  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
4    bounding-box-50  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
5    bounding-box-60  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
6    bounding-box-70  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
7    bounding-box-80  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
8    bounding-box-90  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
9   bounding-box-100  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
10  bounding-box-110  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
11  bounding-box-120  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
12  bounding-box-130  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
13  bounding-box-140  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
14  bounding-box-150  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
15  bounding-box-160  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
16  bounding-box-170  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
17  bounding-box-180  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
18  bounding-box-190  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
19  bounding-box-200  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
20  bounding-box-210  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
21  bounding-box-220  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
22  bounding-box-230  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
23  bounding-box-240  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
24  bounding-box-250  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
25  bounding-box-260  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
26  bounding-box-270  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
27  bounding-box-280  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
28  bounding-box-290  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
29  bounding-box-300  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
..               ...                                                ...   ...   
62  bounding-box-630  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
63  bounding-box-640  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
64  bounding-box-650  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
65  bounding-box-660  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
66  bounding-box-670  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
67  bounding-box-680  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
68  bounding-box-690  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
69  bounding-box-700  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
70  bounding-box-710  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
71  bounding-box-720  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
72  bounding-box-730  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
73  bounding-box-740  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
74  bounding-box-750  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
75  bounding-box-760  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
76  bounding-box-770  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
77  bounding-box-780  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
78  bounding-box-790  https://sierra-videos.s3.amazonaws.com/2017_05...   car   
79  bounding-box-800  https://sierra-videos.

In [47]:
#write results to csv
data.to_csv('bounding-box-results-sample.csv')

## <a id=image-similarity>image-similarity</a>
This API takes in two images, specified by URL, and returns a score for how similar those two images are.  

input: {'input': {'image1': {'url': 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1a.png'}, 'image2': {'url': 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1b.png'}}

result: {'similarityScore': 0.75}

Here's an example of a HIT that's shown to Workers:
![image-similarity-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/image-similarity-HIT-example.png)

In [18]:
function_name = 'image-similarity'

### Create a Task

In [19]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

my-test-task-847ac56e049249228283881984914ebb


In [20]:
# the URLs of the images that you want to compare
image_url_1 = 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1a.png'
image_url_2 = 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1b.png'

In [21]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,task_name,
                                   {'image1': {'url': image_url_1},
                                    'image2': {'url': image_url_2}})

print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))        

PUT response: {'status_code': 201, 'task': {'input': {'image1': {'url': 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1a.png'}, 'image2': {'url': 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1b.png'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-847ac56e049249228283881984914ebb'}}


### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [22]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

GET response: {'status_code': 200, 'task': {'input': {'image1': {'url': 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1a.png'}, 'image2': {'url': 'https://s3-us-west-2.amazonaws.com/mturk-sample-media/images-to-compare/image-similarity-1b.png'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-847ac56e049249228283881984914ebb'}}


## <a id=co-reference-resolution>co-reference-resolution</a>

Co-reference resolution is the task of finding all expressions that refer to the same entity in a text. This API returns references and the position of the start and end of each reference.

input: { "text": "Megan said "I voted for Hilary because she better aligns with my values"" }

result:  {'references': {'ref-1': [{'end': 6, 'start': 0, 'text': 'Megan ', 'type': 'ref-1'}, {'end': 14, 'start': 12, 'text': 'I ', 'type': 'ref-1'}, {'end': 64, 'start': 62, 'text': 'my', 'type': 'ref-1'}], 'ref-2': [{'end': 30, 'start': 24, 'text': 'Hilary', 'type': 'ref-2'}, {'end': 42, 'start': 39, 'text': 'she', 'type': 'ref-2'}]}}

Internal only: this API currently only asks one Worker to complete the HIT, excluding specific Workers who have been put on an exclusion list because of poor past performance on NLP related HITs.

Here's an example of a HIT that's shown to Workers:
![co-reference-resolution-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/co-reference-resolution-HIT-example.png)

### Create a Task

In [ ]:
#set the function_name to the name of the API
function_name = 'co-reference-resolution'

In [ ]:
# automatically generate a random task ID
#task_name = 'my-test-task-' + uuid.uuid4().hex
task_name = 'my-test-task-656e8384878a40d59f55289a48085d0d'
print(task_name)

In [ ]:
# define the text that you want analyzed, up to 16k
text = 'Megan said \"I voted for Hilary because she better aligns with my values\"'

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'text': text})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id=keyphrase-extraction>key-phrase-extraction</a> 

Key phrases provide a concise description of a document’s content; they are useful for document categorization, clustering, indexing, search, and summarization; quantifying semantic similarity with other documents; as well as conceptualizing particular knowledge domains.  Key phrase extraction is the process of extracting the important and topic words in a document.

input: { "text": "This is the second Boots I brought from this brand, the material they are used to make these boots are quite decent, it’s not real leather though, but once you get it on hand you can tell that it is definitely high quality material, the last boots I brought cost me 45 bucks, it lasted about 1 year and half, very remarkable durability I got to say, beside this the interior is super soft and very comfortable to wear. If you are looking for mid-range priced boots with high quality this is the best shot for you." }

result: {'keyPhrases': [{'end': 25, 'start': 19, 'text': 'Boots '}, {'end': 115, 'start': 103, 'text': 'quite decent'}, {'end': 138, 'start': 122, 'text': 'not real leather'}, {'end': 231, 'start': 210, 'text': 'high quality material'}, {'end': 307, 'start': 292, 'text': '1 year and half'}, {'end': 335, 'start': 309, 'text': 'very remarkable durability'}, {'end': 388, 'start': 378, 'text': 'super soft'}, {'end': 409, 'start': 393, 'text': 'very comfortable'}, {'end': 458, 'start': 442, 'text': 'mid-range priced'}, {'end': 482, 'start': 470, 'text': 'high quality'}]}

Internal only: this API currently only asks one Worker to complete the HIT, excluding specific Workers who have been put on an exclusion list because of poor past performance on NLP related HITs.

Here's an example of a HIT that's shown to Workers: 
![key-phrase-extraction-HIT-example](https://s3-us-west-2.amazonaws.com/mturk-sample-tasks/key-phrase-extraction-HIT-example.png)

### Create a Task

In [ ]:
#set the function_name to the name of the API
function_name = 'key-phrase-extraction'

In [ ]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

In [ ]:
# define the text that you want analyzed, up to 16k
text = 'This is the second Boots I brought from this brand, the material they are used to make these boots are quite decent, it’s not real leather though, but once you get it on hand you can tell that it is definitely high quality material, the last boots I brought cost me 45 bucks, it lasted about 1 year and half, very remarkable durability I got to say, beside this the interior is super soft and very comfortable to wear. If you are looking for mid-range priced boots with high quality this is the best shot for you.'

In [ ]:
# create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                             {'text': text})
print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id=face-detection>face-detection</a>

Given an image URL, return bounding boxes around faces in the image.

input: {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/71m-wVEQ9kL._SX522_.jpg'} }

result: {
    'labels': [
      {
        'boundingBox': {
          'height': 115,
          'left': 248,
          'top': 19,
          'width': 116
        }
      },
      {
        'boundingBox': {
          'height': 89,
          'left': 354,
          'top': 45,
          'width': 115
        }
      }
    ]}

TK: include a sample HIT

In [25]:
# TK: function name is probably changing
function_name = 'facial-detection'

### Create a Task

In [21]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

my-test-task-739ac9c6b864406095f6a35145ff0851


In [22]:
# provide the context and the intent
image_url = 'https://images-na.ssl-images-amazon.com/images/I/71m-wVEQ9kL._SX522_.jpg'

In [26]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                            {'image': {'url': image_url}})

print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

PUT response: {'status_code': 201, 'task': {'input': {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/71m-wVEQ9kL._SX522_.jpg'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-739ac9c6b864406095f6a35145ff0851'}}


### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [ ]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

## <a id=logo-recognition>logo-recognition</a>

Given an image URL, return bounding boxes around logos in the image along with the name of the brand it represents.

input: {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'} }

result: 
 {
    'labels': [
      {
        'boundingBox': {
          'height': 35,
          'left': 213,
          'top': 408,
          'width': 98
        },
        'organization': 'amazon.com, inc'
      }
    ]
  }

TK: include a sample HIT

In [31]:
# TK: function name is probably changing
function_name = 'logo-detection'

### Create a Task

In [28]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

my-test-task-f9a3ae43a0ce41ef9998d20f52f6ed6e


In [29]:
# provide the context and the intent
image_url = 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'

In [32]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                            {'image': {'url': image_url}})

print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

PUT response: {'status_code': 201, 'task': {'input': {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-f9a3ae43a0ce41ef9998d20f52f6ed6e'}}


### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [33]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

GET response: {'status_code': 200, 'task': {'input': {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-f9a3ae43a0ce41ef9998d20f52f6ed6e'}}


## <a id=text-recognition>text-recognition</a>

Given an image URL, return the text in the image and the bounding boxes around each word.

input: {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/71m-wVEQ9kL._SX522_.jpg'} }

result: 
 {
    'labels': [
      {
        'boundingBox': {
          'height': 29,
          'left': 71,
          'top': 440,
          'width': 67
        },
        'text': 'room'
      },
      {
        'boundingBox': {
          'height': 33,
          'left': 141,
          'top': 440,
          'width': 65
        },
        'text': 'filling'
      },
      {
        'boundingBox': {
          'height': 28,
          'left': 207,
          'top': 443,
          'width': 71
        },
        'text': 'sound'
      },
      {
        'boundingBox': {
          'height': 28,
          'left': 280,
          'top': 440,
          'width': 46
        },
        'text': 'with'
      },
      {
        'boundingBox': {
          'height': 31,
          'left': 329,
          'top': 441,
          'width': 60
        },
        'text': 'crisp'
      },
      {
        'boundingBox': {
          'height': 22,
          'left': 389,
          'top': 445,
          'width': 69
        },
        'text': 'vocals'
      },
      {
        'boundingBox': {
          'height': 22,
          'left': 117,
          'top': 477,
          'width': 44
        },
        'text': 'and'
      },
      {
        'boundingBox': {
          'height': 23,
          'left': 313,
          'top': 476,
          'width': 99
        },
        'text': 'response'
      },
      {
        'boundingBox': {
          'height': 24,
          'left': 165,
          'top': 476,
          'width': 90
        },
        'text': 'dynamic'
      },
      {
        'boundingBox': {
          'height': 24,
          'left': 259,
          'top': 476,
          'width': 50
        },
        'text': 'bass'
      }
    ]
  }

TK: include a sample HIT

In [38]:
function_name = 'text-recognition'

### Create a Task

In [35]:
# automatically generate a random task ID
task_name = 'my-test-task-' + uuid.uuid4().hex
print(task_name)

my-test-task-6561701d2fa847278d0486a6fb2b370d


In [36]:
# provide the context and the intent
image_url = 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'

In [39]:
#create a single task with the input you specified above
put_result = crowd_client.put_task(function_name,
                             task_name,
                            {'image': {'url': image_url}})

print('PUT response: {}'.format(
    {'status_code': put_result.status_code, 'task': put_result.json()}))

PUT response: {'status_code': 201, 'task': {'input': {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-6561701d2fa847278d0486a6fb2b370d'}}


### Get the result

Wait a few minutes before calling get_task to give Workers a chance to submit answers

In [33]:
get_result = crowd_client.get_task(function_name, task_name)

print('GET response: {}'.format(                     
    {'status_code': get_result.status_code, 'task': get_result.json()}))

GET response: {'status_code': 200, 'task': {'input': {'image': {'url': 'https://images-na.ssl-images-amazon.com/images/I/61yI7vWa83L._SX522_.jpg'}}, 'problemDetails': None, 'result': None, 'state': 'processing', 'taskName': 'my-test-task-f9a3ae43a0ce41ef9998d20f52f6ed6e'}}
